### ISSUE TITLE
Investigate performance of `ConstrainedReschedule` pass

### ISSUE THREAD

**Issue by kdk** [Qiskit/qiskit-terra#10067]
### What should we add?

Following the discussion and investigation of #10049 and #10051 , it came up that there may be some room for improvement in the performance of the `ConstrainedReschedule` pass, but more investigation is needed. This issue is to track what's been looked at so far, and decide how best to resolve.

Loosely, `ConstrainedReschedule` loops over all the nodes in the input DAG to see if any are not aligned with the defined `{acquire,pulse}_alignment` constraints. If any are found, the pass calculates the offset needed to align that node with constraints, and iteratively (or recursively) propagates that shift to any subsequent nodes which depend on the shifted node (and need to have their start time delayed to avoid overlapping).

I ran a few `random_circuit(width=100, depth=_)` circuits through the `PassManager` from #10049 for increasing depths and got the following plot of (scheduling stage) runtime vs. instruction count:
![image](https://user-images.githubusercontent.com/2241698/235798193-8f030308-0fc4-4b81-b494-c7807728b27f.png)

It seems like there are a few worthwhile things to look at:
* Because every gate shift is propagated to every dependent gate, the pass ends up pushing back gates multiple times (and ends up with runtime more cubic ~quadratic~ in the number of instructions, while it seems like ~linear should be possible). Could we instead mark only the set of direct descendants of each node that need to be deferred so that we examine the alignment of each instruction only once? (If we pull the marked descendants out in topological order, can we get the needed shift from the `node_start_time` of that node's predecessors?)

I added a quick change to count the number of times each `node` was included in `nodes_with_overlap` for the example in #10049 (~96k Instructions post-transpile) and got the following map examine counts vs how many nodes were examined that many times. i.e. 556 nodes were examined once, 101 nodes were examined 3M times each, 4 nodes were examined 16 times each, so on.

```python
Counter({1: 556,
         3066550: 84,
         3066549: 17,
         4: 16,
         16: 13,
         68: 12,
         8: 12,
         1002: 11,
         16032: 11,
         4008: 9,
         63: 8,
         251: 8,
         324621: 7,
         17: 6,
         2004: 6,
         32064: 6,
         80154: 6,
         ...
```

* When shifting dependent nodes, we don't currently check that the node we're shifting lands on a valid alignment. Could we check this ahead of time (to avoid the need to re-align this instruction and its dependents later?)
* If we can't improve the performance with either or both of the above, a suggestion from @nkanazawa1989 was to incorporate checking the alignment constraints directly in the `A{SL}APScheduleAnalysis` passes by truncating each calculated start time to the least common multiple of the alignments (this has the disadvantage of mixing the scheduling and alignment concerns, since alignment doesn't apply to all hardware).
